### DISCUSSION TITLE
Issue with Action Grant/Remove

### DISCUSSION THREAD

**Discussion by Sirryan2002** [ParadiseSS13/Paradise#15932] [Code Help] [answered]
Whenever a Species changes it will call [on_species_loss](https://github.com/ParadiseSS13/Paradise/blob/f0841b36294baccc2c5ac18678bd898defb8583d/code/modules/mob/living/carbon/human/species/slime.dm#L71) and [on_species_gain](https://github.com/ParadiseSS13/Paradise/blob/f0841b36294baccc2c5ac18678bd898defb8583d/code/modules/mob/living/carbon/human/species/_species.dm#L281) which is also called in that species DM file.

When ever on_species_lost is called for something like slimes, it remove the actions recolor and regrow limbs
```
/datum/species/slime/on_species_loss(mob/living/carbon/human/H)
	..()
	if(grow)
		grow.Remove(H)
	if(recolor)
		recolor.Remove(H)
	UnregisterSignal(H, COMSIG_HUMAN_UPDATE_DNA)
```
However if you transform back into that species it will call on_species_gain and readd those actions

```
/datum/species/slime/on_species_gain(mob/living/carbon/human/H)
	..()
	grow = new()
	grow.Grant(H)
	recolor = new()
	recolor.Grant(H)
	RegisterSignal(H, COMSIG_HUMAN_UPDATE_DNA, /datum/species/slime/./proc/blend)
	blend(H)
```
So at this point, the mob has both the grow and recolor actions(and the icons), however when the mob transforms back into a different species again, on_species_loss is once again called, HOWEVER, it fails the if checks for both actions even though they were granted back to the mob.
 both `if(grow)` and `if(recolor)` return false and therefor the actions are not removed from the player

I checked the variables and it appears that the mob loses the action `/datum/action/item_action/button` when switching around. If anyone could help me figure out why recolor and grow return false please share your wisdom.

**Comment by SteelSlayer** [comment, marked as answer]
When transforming as a changeling, we end up here:
```dm
/mob/living/carbon/human/proc/change_dna(datum/dna/new_dna, include_species_change = FALSE, keep_flavor_text = FALSE)
	if(include_species_change)
		set_species(new_dna.species.type, retain_damage = TRUE)
	dna = new_dna.Clone()
	[...]
```
The `set_species()` call here  works fine, just as you suspect, the slime gets their `grow` and `recolor` because `on_species_gain()` was called. However the issue happens inside the `Clone()` proc, specifically the line: `new_dna.species = new species.type`

```dm
/datum/dna/proc/Clone()
	var/datum/dna/new_dna = new()
	new_dna.unique_enzymes = unique_enzymes
	new_dna.struc_enzymes_original = struc_enzymes_original // will make clone's SE the same as the original, do we want this?
	new_dna.blood_type = blood_type
	new_dna.real_name = real_name
	new_dna.species = new species.type // fix me
	[...]
```
That line creates a freshly `new`'d species datum which *does not* have any of the code from `on_species_gain()`, because that proc is never ran. 

When we finally return to the `change_dna()` proc, `dna = new_dna.Clone()` resets our species to a blank species datum, erasing the `grow` and `recolor` (or similar) variables.

I *suspect* #15085 was the PR which created this bug.

**Reply by Sirryan2002** [reply]
Alright, this definitely explains it. I'll probably have to called on_species_gain again somewhere to fix this.
